src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Microsoft.WindowsAzure.ResourceStack.Common.Json;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.IO;
+
+namespace Bicep.Core.Emit
+{
+    public class ParametersFileTemplateWriter : TemplateWriter
+    {
+        public ParametersFileTemplateWriter(SemanticModel semanticModel, EmitterSettings settings): base(semanticModel, settings)
+        {
+            this.context = new EmitterContext(semanticModel, settings);
+            this.settings = settings;
+        }
+
+        private readonly EmitterContext context;
+        private readonly EmitterSettings settings;
+
+        public void Write(JsonTextWriter writer, string existingContent)
+        {
+            // Template is used for calcualting template hash, template jtoken is used for writing to file.
+            var templateJToken = GenerateTemplate();
+            if (templateJToken.SelectToken(GeneratorMetadataPath) is not JObject generatorObject)
+            {
+                throw new InvalidOperationException($"generated template doesn't contain a generator object at the path {GeneratorMetadataPath}");
+            }
+
+            if (!string.IsNullOrWhiteSpace(existingContent))
+            {
+                var fileJToken = JToken.Parse(existingContent);
+                var existingParameters = fileJToken.SelectToken("parameters")?.ToObject<JObject>();
+                if (existingParameters is not null)
+                {
+                    var mergeSettings = new JsonMergeSettings { MergeArrayHandling = MergeArrayHandling.Union };
+                    foreach (var existingParameter in existingParameters)
+                    {
+                        var existingParameterName = existingParameter.Key;
+                        var existingParameterValue = existingParameter.Value;
+
+                        if (existingParameterValue is not null)
+                        {
+                            if (templateJToken.SelectToken($"parameters.{existingParameterName}") is JObject existingParameterObject)
```

Comment:
Parameter names (technically *all* property names, but we can probably ignore that for other cases) are case-insensitive, so you need to match case-insensitively.